File: src/main/java/com/social/meow/service/PostService.java
package com.social.meow.service;

import com.social.meow.model.Post;
import com.social.meow.repository.PostRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.dao.DataAccessException;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.redis.core.*;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class PostService {

    private final PostRepository postRepository;
    private final RedisTemplate<String, Object> redisTemplate;
    private final CacheManager cacheManager;

    @Value("${timeline.cache.key}")
    private String timelineCacheKey;

    @Value("${post.cache.key}")
    private String postCacheKey;

    @Value("${timeline.cache.maxSize}")
    private int maxTimelineSize;

    public PostService(PostRepository postRepository, RedisTemplate<String, Object> redisTemplate, CacheManager cacheManager) {
        this.postRepository = postRepository;
        this.redisTemplate = redisTemplate;
        this.cacheManager = cacheManager;
    }

    public List<Post> getAllPosts() {
        return postRepository.findAll();
    }

    public Post getPostById(long id) {
        Optional<Post> postData = postRepository.findById(id);
        return postData.orElse(null);
    }

    public Post createPost(Post post) {
        Post savedPost = postRepository.save(post);
        redisTemplate.executePipelined(new SessionCallback<Object>() {
            @Override
            public <K, V> Object execute(RedisOperations<K, V> operations) throws DataAccessException {
                RedisOperations<String, Object> opsTemplate = (RedisOperations<String, Object>) operations;
                ListOperations<String, Object> listOps = opsTemplate.opsForList();

                listOps.leftPush(timelineCacheKey, savedPost.getId());
                listOps.trim(timelineCacheKey, 0, maxTimelineSize - 1);

                return null;
            }
        });

        return savedPost;
    }

    public Post updatePost(long id, Post post) {
        Optional<Post> postData = postRepository.findById(id);
        if (postData.isPresent()) {
            Post existingPost = postData.get();
            existingPost.setText(post.getText());
            existingPost.setUserId(post.getUserId());

            Post savedPost = postRepository.save(existingPost);
            Cache postsCache = cacheManager.getCache(postCacheKey);
            if (postsCache != null) {
                postsCache.evict(savedPost.getId());
            }
            return savedPost;
        }
        return null;
    }

    public void deletePost(long id) {
        postRepository.deleteById(id);
        Cache postsCache = cacheManager.getCache(postCacheKey);
        if (postsCache != null) {
            postsCache.evict(id);
        }
        redisTemplate.executePipelined(new SessionCallback<Object>() {
            @Override
            public <K, V> Object execute(RedisOperations<K, V> operations) throws DataAccessException {
                RedisOperations<String, Object> opsTemplate = (RedisOperations<String, Object>) operations;
                ListOperations<String, Object> listOps = opsTemplate.opsForList();

                listOps.remove(timelineCacheKey, 0, id);

                return null;
            }
        });
    }

    public List<Post> getPostsByLastIdAndPageSize(Long lastId, int pageSize) {
        Pageable pageable = PageRequest.of(0, pageSize, Sort.by("id").descending());
        Cache postsCache = cacheManager.getCache(postCacheKey);
        List<Post> posts;
        if (lastId == null) {
            posts = postRepository.findAll(pageable).getContent();
        }
        else {
            posts = postRepository.findByIdLessThanOrderByIdDesc(lastId, pageable);
        }
        if (postsCache != null) {
            for (Post post : posts) {
                postsCache.put(post.getId(), post);
            }
        }
        return posts;
    }

    public List<Post> getPostsByIds(List<Long> ids) {
        if (ids == null || ids.isEmpty()) {
            return Collections.emptyList();
        }

        Cache postsCache = cacheManager.getCache(postCacheKey);
        Map<Long, Post> result = new LinkedHashMap<>();
        List<Long> missedIds = new ArrayList<>();

        for (Long id : ids) {
            Post cached = postsCache != null ? postsCache.get(id, Post.class) : null;
            if (cached != null) {
                result.put(id, cached);
            } else {
                missedIds.add(id);
            }
        }

        if (!missedIds.isEmpty()) {
            List<Post> dbPosts = postRepository.findAllById(missedIds);

            if (postsCache != null) {
                for (Post post : dbPosts) {
                    postsCache.put(post.getId(), post);
                    result.put(post.getId(), post);
                }
            } else {
                dbPosts.forEach(p -> result.put(p.getId(), p));
            }
        }

        return ids.stream()
                .map(result::get)
                .filter(Objects::nonNull)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/social/meowkotlin/service/PostService.kt
package com.social.meowkotlin.service

import com.social.meowkotlin.model.Post
import com.social.meowkotlin.repository.PostRepository
import org.springframework.beans.factory.annotation.Value
import org.springframework.cache.Cache
import org.springframework.cache.CacheManager
import org.springframework.dao.DataAccessException
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.redis.core.ListOperations
import org.springframework.data.redis.core.RedisOperations
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.core.SessionCallback
import org.springframework.stereotype.Service
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors

@Service
class PostService(
    private val postRepository: PostRepository,
    private val redisTemplate: RedisTemplate<String, Any>,
    private val cacheManager: CacheManager
) {

    @Value("\${timeline.cache.key}")
    private val timelineCacheKey: String? = null

    @Value("\${post.cache.key}")
    private val postCacheKey: String? = null

    @Value("\${timeline.cache.maxSize}")
    private val maxTimelineSize = 0

    fun getPostById(id: Long): Post? {
        val postData: Optional<Post?> = postRepository.findById(id)
        return postData.orElse(null)
    }

    fun createPost(post: Post): Post {
        val savedPost: Post = postRepository.save(post)
        redisTemplate.executePipelined(object : SessionCallback<Any?> {
            @Throws(DataAccessException::class)
            override fun <K, V> execute(operations: RedisOperations<K, V>): Any? {
                val opsTemplate: RedisOperations<String, Any> = operations as RedisOperations<String, Any>
                val listOps: ListOperations<String, Any> = opsTemplate.opsForList()

                if (timelineCacheKey !== null) {
                    listOps.leftPush(timelineCacheKey, savedPost.id)
                    listOps.trim(timelineCacheKey, 0, (maxTimelineSize - 1).toLong())
                }

                return null
            }
        })

        return savedPost
    }

    fun updatePost(id: Long, post: Post): Post? {
        val postData: Optional<Post?> = postRepository.findById(id)
        if (postData.isPresent()) {
            val existingPost: Post = postData.get()
            existingPost.text = post.text
            existingPost.userId = post.userId

            val savedPost: Post = postRepository.save(existingPost)
            if (postCacheKey !== null) {
                val postsCache: Cache? = cacheManager.getCache(postCacheKey)
                postsCache?.evict(savedPost.id)
            }
            return savedPost
        }
        return null
    }

    fun deletePost(id: Long) {
        postRepository.deleteById(id)
        if (postCacheKey !== null) {
            val postsCache: Cache? = cacheManager.getCache(postCacheKey)
            postsCache?.evict(id)
        }
        redisTemplate.executePipelined(object : SessionCallback<Any?> {
            @Throws(DataAccessException::class)
            override fun <K, V> execute(operations: RedisOperations<K, V>): Any? {
                val opsTemplate: RedisOperations<String, Any> = operations as RedisOperations<String, Any>
                val listOps: ListOperations<String, Any> = opsTemplate.opsForList()

                if (timelineCacheKey !== null) listOps.remove(timelineCacheKey, 0, id)

                return null
            }
        })
    }

    fun getPostsByLastIdAndPageSize(lastId: Long?, pageSize: Int): List<Post> {
        val pageable: Pageable = PageRequest.of(0, pageSize, Sort.by("id").descending())
        val posts: List<Post> = if (lastId == null) {
            postRepository.findAll(pageable).getContent()
        } else {
            postRepository.findByIdLessThanOrderByIdDesc(lastId, pageable)
        }
        if (postCacheKey !== null) {
            val postsCache: Cache? = cacheManager.getCache(postCacheKey)
            if (postsCache != null) {
                for (post in posts) {
                    postsCache.put(post.id, post)
                }
            }
        }
        return posts
    }

    fun getPostsByIds(ids: List<Long>): List<Post> {
        if (ids == null || ids.isEmpty()) {
            return emptyList<Post>()
        }

        val postsCache: Cache? = postCacheKey?.let { cacheManager.getCache(it) }
        val result: MutableMap<Long, Post> = LinkedHashMap<Long, Post>()
        val missedIds: MutableList<Long?> = ArrayList()

        for (id in ids) {
            val cached: Post? = if (postsCache != null) postsCache.get(id, Post::class.java) else null
            if (cached != null) {
                result[id] = cached
            } else {
                missedIds.add(id)
            }
        }

        if (missedIds.isNotEmpty()) {
            val dbPosts: List<Post> = postRepository.findAllById(missedIds)

            if (postsCache != null) {
                for (post in dbPosts) {
                    postsCache.put(post.id, post)
                    result[post.id] = post
                }
            } else {
                dbPosts.forEach(Consumer { p: Post -> result[p.id] = p })
            }
        }

        return ids.mapNotNull { result[it] }
    }
}
